rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -109,21 +98,15 @@ class Publisher : public PublisherBase
     // interprocess publish, resulting in lower publish-to-subscribe latency.
     // It's not possible to do that with an unique_ptr,
     // as do_intra_process_publish takes the ownership of the message.
-    uint64_t message_seq;
     bool inter_process_publish_needed =
       get_subscription_count() > get_intra_process_subscription_count();
-    MessageSharedPtr shared_msg;
+
     if (inter_process_publish_needed) {
-      shared_msg = std::move(msg);
-      message_seq =
-        store_intra_process_message(intra_process_publisher_id_, shared_msg);
+      std::shared_ptr<MessageT> shared_msg = std::move(msg);
+      this->do_intra_process_publish(shared_msg);
```

Comment:
This will require to have some of the logic of intra-process communication also in the `publisher.hpp` file instead of only in intra-process-manager.
Is it worth?

Reply:
Sorry, forgot to come back to this one.

Yes, I think it's worth it, but maybe we could reorganize the API to avoid more coupling.

Right now, I believe due to this, the tutorials behind the `intra_process_demos` package no longer work correctly. As they expect the same pointer to be used iteration to iteration.

Reply:
I changed the behavior in 3013e02, please have a look if you have time. I'll  probably push ahead with that change for now, but I can fix it up later if you think it needs it.

With that change the `cyclic_pipeline` demo works again.